jupyter-extension/jupyterlab_pachyderm/pps_client.py

```diff
@@ -279,32 +175,150 @@ def entrypoint():
+    async def generate(path):
+        """Generates the pipeline spec from the Notebook file specified.
+
+        Args:
+            path: The path (within Jupyter) to the Notebook file.
+        """
+        get_logger().debug(f"path: {path}")
+
+        path = Path(path.lstrip("/"))
+        if not path.exists():
+            raise HTTPError(status_code=400, reason=f"notebook does not exist: {path}")
+
+        try:
+            config = PpsConfig.from_notebook(path)
+        except ValueError as err:
+            raise HTTPError(status_code=400, reason=str(err))
+
+        pipeline_spec = create_pipeline_spec(config, "...")
+        return pipeline_spec.to_json()
+
+    async def create(self, path: str, body: dict):
+        """Creates the pipeline from the Notebook file specified.
+
+        Args:
+            path: The path (within Jupyter) to the Notebook file.
+            body: The body of the request.
+        """
+        get_logger().debug(f"path: {path} | body: {body}")
+
+        path = Path(path.lstrip("/"))
+        if not path.exists():
+            raise HTTPError(status_code=400, reason=f"notebook does not exist: {path}")
+
+        check = body.get("last_modified_time")
+        if not check:
+            raise HTTPError(
+                status_code=400, reason="Bad Request: last_modified_time not specified"
+            )
+        check = datetime.fromisoformat(check.rstrip("Z"))
+
+        last_modified = datetime.utcfromtimestamp(os.path.getmtime(path))
+        if check != last_modified:
+            raise HTTPError(
+                status_code=400,
+                reason=f"stale notebook: client time ({check}) != on-disk time ({last_modified})",
+            )
+        try:
+            config = PpsConfig.from_notebook(path)
+        except ValueError as err:
+            raise HTTPError(status_code=400, reason=str(err))
 
-    return branch_name
+        if config.requirements and not os.path.exists(config.requirements):
+            raise HTTPError(status_code=400, reason="requirements file does not exist")
+
+        script, _resources = self.nbconvert.from_filename(str(path))
+
+        companion_repo = pfs.Repo(
+            name=companion_repo_name(config),
+            project=pfs.Project(name=config.pipeline.project.name),
+        )
+        for r in self.client.pfs.list_repo():
+            same_project = config.pipeline.project.name == r.repo.project.name
+            if (companion_repo.name == r.repo.name) and same_project:
+                break
+        else:
+            self.client.pfs.create_repo(
+                repo=companion_repo,
+                description=f"files for running the {config.pipeline.name} pipeline",
+            )
+
+        companion_branch = upload_environment(
+            self.client, companion_repo, config, script.encode("utf-8")
+        )
+        pipeline_spec = create_pipeline_spec(config, companion_branch)
+        try:
+            self.client.pps.create_pipeline(
```

Comment:
Well at the very least `CreatePipelineV2` would make this code better. The other options is to pass `CreatePipelineRequest` message to the `self.client.pps._ApiStub__rpc_create_pipeline` but I'm not sure that's any better.